server/util/rexec/rexec.go

```diff
@@ -200,6 +202,55 @@ func Prepare(ctx context.Context, env environment.Env, instanceName string, dige
 			return nil, err
 		}
 		inputRootDigest = d
+	} else {
+		eg.Go(func() error {
+			casClient := env.GetContentAddressableStorageClient()
+			// Call GetTree to ensure that the input root directory is fully
+			// uploaded to the CAS, even if it is empty.
+			treeStream, err := casClient.GetTree(ctx, &repb.GetTreeRequest{
+				InstanceName:   instanceName,
+				RootDigest:     action.GetInputRootDigest(),
+				DigestFunction: digestFunction,
+			})
+			if err != nil {
+				return status.InternalErrorf("failed to get tree for input root digest %q: %s", action.GetInputRootDigest(), err)
+			}
+			var blobDigests []*repb.Digest
+			for {
+				treeResp, err := treeStream.Recv()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return status.InternalErrorf("failed to receive tree response for input root digest %q: %s", action.GetInputRootDigest(), err)
+				}
+				for _, d := range treeResp.GetDirectories() {
+					for _, fn := range d.GetFiles() {
+						blobDigests = append(blobDigests, fn.GetDigest())
+					}
+				}
+			}
+			if len(blobDigests) == 0 {
+				return nil
+			}
+			blobDigests = slices.Clip(blobDigests)
+			missingBlobResp, err := casClient.FindMissingBlobs(ctx, &repb.FindMissingBlobsRequest{
+				InstanceName:   instanceName,
+				DigestFunction: digestFunction,
+				BlobDigests:    blobDigests,
+			})
+			if err != nil {
+				return status.InternalErrorf("failed to find missing blobs for input root digest %q: %s", action.GetInputRootDigest(), err)
+			}
+			if len(missingBlobResp.GetMissingBlobDigests()) == 0 {
+				return nil
+			}
+			missingBlobJson, err := protojson.Marshal(missingBlobResp)
+			if err != nil {
+				return status.InternalErrorf("failed to marshal missing blob response: %s", err)
+			}
+			return status.FailedPreconditionErrorf("input root digest %q is not fully uploaded to the CAS, missing blobs: %s", action.GetInputRootDigest(), missingBlobJson)
+		})
```

Comment:
I wonder if it would be faster to upload a fake action result with the input root tree as one of the output directories, and then request for that action result.

At least it would be lower in terms of bytes transferred. 🤔